src/Npgsql/NpgsqlStatement.cs

```diff
@@ -105,8 +114,12 @@ internal void Reset()
             _description = null;
             LongRows = 0;
             OID = 0;
-            InputParameters.Clear();
             PreparedStatement = null;
+
+            if (ReferenceEquals(_inputParameters, _ownedInputParameters))
+                InputParameters.Clear();
+            else if (_inputParameters is not null)
+                _inputParameters = null; // We're pointing at a user's NpgsqlParameterCollection
```

Comment:
Can you provide a problematic scenario?

The idea is that as long as we used named parameters, `_inputParameters == _ownedInputParameters` (we have to rearrange the parameters in the right order). Once a positional command is executed, _inputParameters now points directly to the user's NpgsqlParameterCollection (no need to rearrange). Resetting at that point makes it null, so the next time it's accessed it will point to _ownedInputParameters again.

Reply:
> Can you provide a problematic scenario?

1. Get the `InputParameters` (so both `_inputParameters` and `_ownedInputParameters` reference the same list).
2. Add a param to the `InputParameters`.
3. Set the `InputParameters` with a new list.
4. After a statement is reset, the param from the second step should leak.

Reply:
Before step 3, we should always be calling Reset (i.e. from TruncateStatementsInOne), which would should Clear those lists, no?

Reply:
Just as an example:

```csharp
using var conn1 = OpenConnection();
var preparedStatement = "";
using (var cmd1 = new NpgsqlCommand("SELECT @p", conn1))
{
	cmd1.Parameters.AddWithValue("@p", 42);
	cmd1.Prepare();
	cmd1.Statements[0].InputParameters = new List<NpgsqlParameter>(new[] { new NpgsqlParameter("@p", 8) });
	cmd1.CommandText = "SELECT 1, @p";
	cmd1.Prepare();
	Assert.AreEqual(1, cmd1.Statements[0].InputParameters.Count);
}
```

Reply:
As discussed offline, the setter on InputParameters is internal - I'll add an assertion to be on the safe side though.